File: src/sync.cpp
#include "nso_album_sync/sync.hpp"

#include "nso_album_sync/game_aliases.hpp"
#include "nso_album_sync/util.hpp"

#ifdef _WIN32
#include "nso_album_sync/windows_compat.hpp"
#endif

#include <chrono>
#include <cctype>
#include <cstdint>
#include <ctime>
#include <fstream>
#include <set>
#include <utility>
#include <stdexcept>
#include <unordered_map>
#include <vector>

namespace nso {
namespace {

struct ExistingAlbumIndex {
    std::set<std::string> filenames_and_prefixes;
    std::unordered_map<std::string, std::string> folder_by_timestamp_prefix;
};

bool should_cancel(const std::function<bool()>& cancelled) {
    return cancelled && cancelled();
}

void throw_if_cancelled(const std::function<bool()>& cancelled) {
    if (should_cancel(cancelled)) throw std::runtime_error("Sync cancelled");
}

void append_utf8(std::string& output, std::uint32_t code_point) {
    if (code_point <= 0x7f) {
        output.push_back(static_cast<char>(code_point));
    } else if (code_point <= 0x7ff) {
        output.push_back(static_cast<char>(0xc0 | (code_point >> 6)));
        output.push_back(static_cast<char>(0x80 | (code_point & 0x3f)));
    } else if (code_point <= 0xffff) {
        output.push_back(static_cast<char>(0xe0 | (code_point >> 12)));
        output.push_back(static_cast<char>(0x80 | ((code_point >> 6) & 0x3f)));
        output.push_back(static_cast<char>(0x80 | (code_point & 0x3f)));
    } else {
        output.push_back(static_cast<char>(0xf0 | (code_point >> 18)));
        output.push_back(static_cast<char>(0x80 | ((code_point >> 12) & 0x3f)));
        output.push_back(static_cast<char>(0x80 | ((code_point >> 6) & 0x3f)));
        output.push_back(static_cast<char>(0x80 | (code_point & 0x3f)));
    }
}

std::pair<std::uint32_t, std::size_t> decode_utf8(
    const std::string& text,
    std::size_t offset) {
    const auto first = static_cast<unsigned char>(text[offset]);
    if (first < 0x80) return {first, 1};

    auto continuation = [&](std::size_t index) -> int {
        if (index >= text.size()) return -1;
        const auto byte = static_cast<unsigned char>(text[index]);
        return (byte & 0xc0) == 0x80 ? (byte & 0x3f) : -1;
    };

    if ((first & 0xe0) == 0xc0) {
        const int b1 = continuation(offset + 1);
        if (b1 >= 0) return {((first & 0x1fU) << 6U) | static_cast<unsigned>(b1), 2};
    } else if ((first & 0xf0) == 0xe0) {
        const int b1 = continuation(offset + 1);
        const int b2 = continuation(offset + 2);
        if (b1 >= 0 && b2 >= 0) {
            return {((first & 0x0fU) << 12U) |
                        (static_cast<unsigned>(b1) << 6U) |
                        static_cast<unsigned>(b2),
                    3};
        }
    } else if ((first & 0xf8) == 0xf0) {
        const int b1 = continuation(offset + 1);
        const int b2 = continuation(offset + 2);
        const int b3 = continuation(offset + 3);
        if (b1 >= 0 && b2 >= 0 && b3 >= 0) {
            return {((first & 0x07U) << 18U) |
                        (static_cast<unsigned>(b1) << 12U) |
                        (static_cast<unsigned>(b2) << 6U) |
                        static_cast<unsigned>(b3),
                    4};
        }
    }
    return {first, 1};
}

std::string normalize_for_matching_v1(const std::string& text) {
    std::string normalized;
    normalized.reserve(text.size());

    for (std::size_t offset = 0; offset < text.size();) {
        const auto [code_point, width] = decode_utf8(text, offset);
        offset += width;

        if (code_point >= 0x0300 && code_point <= 0x036f) continue;

        if (code_point < 0x80) {
            const auto character = static_cast<unsigned char>(code_point);
            if (std::isalnum(character)) {
                normalized.push_back(static_cast<char>(std::tolower(character)));
            }
            continue;
        }

        char latin = '\0';
        switch (code_point) {
            case 0x00c0: case 0x00c1: case 0x00c2: case 0x00c3: case 0x00c4: case 0x00c5:
            case 0x00e0: case 0x00e1: case 0x00e2: case 0x00e3: case 0x00e4: case 0x00e5:
                latin = 'a'; break;
            case 0x00c7: case 0x00e7: latin = 'c'; break;
            case 0x00c8: case 0x00c9: case 0x00ca: case 0x00cb:
            case 0x00e8: case 0x00e9: case 0x00ea: case 0x00eb:
                latin = 'e'; break;
            case 0x00cc: case 0x00cd: case 0x00ce: case 0x00cf:
            case 0x00ec: case 0x00ed: case 0x00ee: case 0x00ef:
                latin = 'i'; break;
            case 0x00d1: case 0x00f1: latin = 'n'; break;
            case 0x00d2: case 0x00d3: case 0x00d4: case 0x00d5: case 0x00d6: case 0x00d8:
            case 0x00f2: case 0x00f3: case 0x00f4: case 0x00f5: case 0x00f6: case 0x00f8:
                latin = 'o'; break;
            case 0x00d9: case 0x00da: case 0x00db: case 0x00dc:
            case 0x00f9: case 0x00fa: case 0x00fb: case 0x00fc:
                latin = 'u'; break;
            case 0x00dd: case 0x0178: case 0x00fd: case 0x00ff:
                latin = 'y'; break;
            default: break;
        }
        if (latin != '\0') {
            normalized.push_back(latin);
            continue;
        }

        // .NET v1 kept non-Latin letters/digits after Unicode decomposition.
        // Preserve their UTF-8 representation so exact Japanese and other
        // localized aliases continue to compare instead of collapsing to empty.
        append_utf8(normalized, code_point);
    }
    return normalized;
}

std::string sanitize_folder_v1(const std::string& text) {
    if (trim(text).empty()) return "Other";
    std::string clean;
    clean.reserve(text.size());
    for (const unsigned char character : text) {
        if (character < 0x20) continue;
        switch (character) {
            case '<': case '>': case ':': case '"': case '/':
            case '\\': case '|': case '?': case '*':
                continue;
            default:
                clean.push_back(static_cast<char>(character));
                break;
        }
    }
    clean = trim(std::move(clean));
    return clean.empty() ? "Other" : clean;
}

bool equal_v1_name(const std::string& left, const std::string& right) {
    return lower(left) == lower(right);
}

std::string capture_timestamp_prefix(std::int64_t timestamp) {
    if (timestamp <= 0) timestamp = unix_now();
    const auto seconds = timestamp > 10'000'000'000LL ? timestamp / 1000 : timestamp;
    const auto time = static_cast<std::time_t>(seconds);
    std::tm local_time{};
#ifdef _WIN32
    localtime_s(&local_time, &time);
#else
    localtime_r(&time, &local_time);
#endif
    char buffer[32];
    std::strftime(buffer, sizeof(buffer), "%Y%m%d%H%M%S", &local_time);
    return std::string(buffer) + "00";
}

std::string prefix_from_existing_filename(const std::filesystem::path& path) {
    std::string prefix = path.stem().string();
    if (prefix.size() > 2 && prefix.ends_with("_c")) {
        prefix.resize(prefix.size() - 2);
    } else if (prefix.size() > 3 && prefix.ends_with("-00")) {
        prefix.resize(prefix.size() - 3);
    }
    return prefix;
}

ExistingAlbumIndex index_existing_album(
    const std::filesystem::path& root,
    const std::function<bool()>& cancelled) {
    ExistingAlbumIndex index;
    for (auto iterator = std::filesystem::recursive_directory_iterator(
             root, std::filesystem::directory_options::skip_permission_denied);
         iterator != std::filesystem::recursive_directory_iterator();
         ++iterator) {
        throw_if_cancelled(cancelled);
        if (!iterator->is_regular_file()) continue;

        const auto filename_lower = lower(iterator->path().filename().string());
        if (filename_lower.ends_with(".part") || filename_lower.ends_with(".tmp")) {
            continue;
        }

        std::error_code error;
        if (iterator->file_size(error) == 0) continue;
        const auto& path = iterator->path();
        const auto filename = path.filename().string();
        const auto prefix = prefix_from_existing_filename(path);
        index.filenames_and_prefixes.insert(lower(filename));
        index.filenames_and_prefixes.insert(lower(prefix));
        index.folder_by_timestamp_prefix.emplace(
            lower(prefix), path.parent_path().filename().string());
    }
    return index;
}

std::unordered_map<std::string, std::string> learn_title_folders(
    const std::vector<MediaItem>& media,
    const ExistingAlbumIndex& existing) {
    std::unordered_map<std::string, std::string> folders;
    for (const auto& item : media) {
        if (item.title_id.empty()) continue;
        const auto timestamp = item.captured_at != 0 ? item.captured_at : item.uploaded_at;
        const auto prefix = capture_timestamp_prefix(timestamp);
        const auto existing_folder =
            existing.folder_by_timestamp_prefix.find(lower(prefix));
        if (existing_folder != existing.folder_by_timestamp_prefix.end()) {
            folders[lower(item.title_id)] = existing_folder->second;
        }
    }
    return folders;
}

void preserve_capture_timestamp(
    const std::filesystem::path& path,
    std::int64_t timestamp) {
    if (timestamp <= 0) return;
    try {
        const auto seconds_value =
            timestamp > 10'000'000'000LL ? timestamp / 1000 : timestamp;
        const auto seconds = std::chrono::seconds(seconds_value);
        const auto capture_time = std::chrono::system_clock::time_point(seconds);
#ifdef _WIN32
        constexpr std::uint64_t kUnixToWindowsEpochSeconds = 11'644'473'600ULL;
        if (seconds_value >= 0) {
            ULARGE_INTEGER raw_time{};
            raw_time.QuadPart =
                (static_cast<std::uint64_t>(seconds_value) +
                 kUnixToWindowsEpochSeconds) * 10'000'000ULL;
            FILETIME file_time{};
            file_time.dwLowDateTime = raw_time.LowPart;
            file_time.dwHighDateTime = raw_time.HighPart;
            HANDLE file = CreateFileW(
                path.c_str(), FILE_WRITE_ATTRIBUTES,
                FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
                nullptr, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, nullptr);
            if (file != INVALID_HANDLE_VALUE) {
                SetFileTime(file, &file_time, nullptr, &file_time);
                CloseHandle(file);
            }
        }
#endif
        const auto file_time =
            std::filesystem::file_time_type::clock::now() +
            (capture_time - std::chrono::system_clock::now());
        std::filesystem::last_write_time(path, file_time);
    } catch (...) {
    }
}

}  // namespace

std::string SyncEngine::resolve_game_folder(
    const std::filesystem::path& album_directory,
    const std::string& app_name) const {
    const auto default_clean = sanitize_folder_v1(app_name);
    if (!std::filesystem::exists(album_directory) || trim(app_name).empty()) {
        return default_clean;
    }

    try {
        struct DirectoryName {
            std::string original;
            std::string normalized;
        };
        std::vector<DirectoryName> directories;
        for (const auto& directory : std::filesystem::directory_iterator(album_directory)) {
            if (!directory.is_directory()) continue;
            const auto name = directory.path().filename().string();
            directories.push_back({name, normalize_for_matching_v1(name)});
        }

        const auto normalized_app_name = normalize_for_matching_v1(app_name);
        std::vector<std::string> synonyms{app_name};
        for (const auto& group : game_alias_groups()) {
            bool group_matches = false;
            for (const auto& alias : group) {
                if (normalize_for_matching_v1(alias) == normalized_app_name) {
                    group_matches = true;
                    break;
                }
            }
            if (group_matches) {
                synonyms = group;
                break;
            }
        }

        // Match v1.0.0 ordering: sanitized exact match first, then normalized
        // exact match for every known localization/synonym.
        for (const auto& synonym : synonyms) {
            const auto clean_synonym = sanitize_folder_v1(synonym);
            const auto normalized_synonym = normalize_for_matching_v1(synonym);

            for (const auto& directory : directories) {
                if (equal_v1_name(directory.original, clean_synonym)) {
                    return directory.original;
                }
            }
            for (const auto& directory : directories) {
                if (directory.normalized == normalized_synonym) {
                    return directory.original;
                }
            }
        }

        // v1's final fallback only fuzzed the API title itself, not every alias.
        if (!normalized_app_name.empty()) {
            for (const auto& directory : directories) {
                if (directory.normalized.size() >= 6 &&
                    normalized_app_name.size() >= 6 &&
                    (directory.normalized.find(normalized_app_name) != std::string::npos ||
                     normalized_app_name.find(directory.normalized) != std::string::npos)) {
                    return directory.original;
                }
            }
        }
    } catch (...) {
        // v1 treated local folder enumeration/matching failures as non-fatal and
        // simply fell back to the sanitized API title.
    }

    return default_clean;
}

SyncResult SyncEngine::sync(const std::function<bool()>& cancelled) {
    const auto config = config_.snapshot();
    if (config.session_token.empty()) {
        throw std::runtime_error("Not signed in to Nintendo Account");
    }

    throw_if_cancelled(cancelled);
    const auto media = coral_.media_list(config.session_token);
    throw_if_cancelled(cancelled);

    const std::filesystem::path root = config.destination_folder.empty()
        ? std::filesystem::current_path() / "Nintendo Switch Album"
        : std::filesystem::path(config.destination_folder);
    std::filesystem::create_directories(root);

    auto existing = index_existing_album(root, cancelled);
    const auto title_folders = learn_title_folders(media, existing);
    const bool root_is_album_directory = lower(root.filename().string()) == "album";
    const auto album_directory = root_is_album_directory ? root : root / "Album";

    int downloaded = 0;
    for (const auto& item : media) {
        throw_if_cancelled(cancelled);
        const auto timestamp = item.captured_at != 0 ? item.captured_at : item.uploaded_at;
        const auto prefix = capture_timestamp_prefix(timestamp);
        const auto extension = lower(item.type) == "video" ? "mp4" : "jpg";
        const auto filename = prefix + "_c." + extension;

        if (existing.filenames_and_prefixes.contains(lower(filename)) ||
            existing.filenames_and_prefixes.contains(lower(prefix))) {
            continue;
        }

        std::string game_folder;
        if (const auto known_folder = title_folders.find(lower(item.title_id));
            known_folder != title_folders.end()) {
            game_folder = known_folder->second;
        } else {
            game_folder = resolve_game_folder(album_directory, item.app_name);
        }

        const auto destination = album_directory / game_folder / filename;
        std::filesystem::create_directories(destination.parent_path());

        const auto response = http_.get(item.content_uri, {}, 60);
        if (response.status / 100 != 2) {
            throw std::runtime_error(
                "Media download failed (HTTP " + std::to_string(response.status) + ")");
        }
        throw_if_cancelled(cancelled);

        auto temporary = destination;
        temporary += ".part";
        std::error_code cleanup_error;
        std::filesystem::remove(temporary, cleanup_error);

        try {
            {
                std::ofstream file(temporary, std::ios::binary | std::ios::trunc);
                if (!file) throw std::runtime_error("Could not create media file");
                file.write(
                    reinterpret_cast<const char*>(response.body.data()),
                    static_cast<std::streamsize>(response.body.size()));
                file.flush();
                if (!file) throw std::runtime_error("Could not write media file");
            }
            throw_if_cancelled(cancelled);

            std::error_code rename_error;
            std::filesystem::rename(temporary, destination, rename_error);
            if (rename_error) {
                throw std::runtime_error(
                    "Could not finalize downloaded media: " + rename_error.message());
            }
        } catch (...) {
            std::error_code ignored;
            std::filesystem::remove(temporary, ignored);
            throw;
        }

        preserve_capture_timestamp(destination, timestamp);
        existing.filenames_and_prefixes.insert(lower(filename));
        existing.filenames_and_prefixes.insert(lower(prefix));
        ++downloaded;
    }

    return {static_cast<int>(media.size()), downloaded};
}

}  // namespace nso

File: src/sync.rs
//! Album synchronization and v1-compatible folder matching.

use crate::config::{ConfigManager, default_album_folder};
use crate::coral::CoralClient;
use crate::game_aliases::GAME_ALIAS_GROUPS;
use crate::http::HttpClient;
use crate::model::{MediaItem, SyncResult};
use filetime::{FileTime, set_file_mtime};
use std::collections::{HashMap, HashSet};
use std::fs::{self, File};
use std::io::Write as _;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use unicode_normalization::{UnicodeNormalization as _, char::is_combining_mark};
use walkdir::WalkDir;

const MAX_MEDIA_DOWNLOAD_BYTES: usize = 256 * 1024 * 1024;

pub struct SyncEngine {
    config: ConfigManager,
    coral: Arc<CoralClient>,
    http: HttpClient,
}

impl SyncEngine {
    pub fn new(config: ConfigManager, coral: Arc<CoralClient>, http: HttpClient) -> Self {
        Self { config, coral, http }
    }

    pub fn sync<F>(&self, cancelled: F) -> anyhow::Result<SyncResult>
    where
        F: Fn() -> bool,
    {
        let config = self.config.snapshot();
        anyhow::ensure!(
            !config.session_token().is_empty(),
            "Not signed in to Nintendo Account"
        );
        check_cancelled(&cancelled)?;
        let media = self.coral.media_list(config.session_token())?;
        check_cancelled(&cancelled)?;

        let root = if config.destination_folder().trim().is_empty() {
            default_album_folder()
        } else {
            PathBuf::from(config.destination_folder())
        };
        fs::create_dir_all(&root)?;
        let mut existing = index_existing_album(&root, &cancelled)?;
        let title_folders = learn_title_folders(&media, &existing);
        let root_is_album = root
            .file_name()
            .and_then(|name| name.to_str())
            .is_some_and(|name| name.eq_ignore_ascii_case("album"));
        let album_directory = if root_is_album {
            root.clone()
        } else {
            root.join("Album")
        };

        let mut downloaded = 0_usize;
        for item in &media {
            check_cancelled(&cancelled)?;
            let timestamp = item.timestamp();
            let prefix = capture_timestamp_prefix(timestamp);
            let extension = if item.media_type().eq_ignore_ascii_case("video") {
                "mp4"
            } else {
                "jpg"
            };
            let filename = format!("{prefix}_c.{extension}");
            if existing.names.contains(&filename.to_ascii_lowercase())
                || existing.names.contains(&prefix.to_ascii_lowercase())
            {
                continue;
            }
            validate_media_item(item)?;

            let game_folder = title_folders
                .get(&item.title_id().to_ascii_lowercase())
                .cloned()
                .unwrap_or_else(|| resolve_game_folder(&album_directory, item.app_name()));
            let destination = album_directory.join(game_folder).join(&filename);
            if let Some(parent) = destination.parent() {
                fs::create_dir_all(parent)?;
            }

            // Redirects are disabled for media. Validation therefore applies to
            // the actual host that receives the request, not just the first URL.
            let response = self.http.get_no_redirect(
                item.content_uri(),
                &[],
                60,
                MAX_MEDIA_DOWNLOAD_BYTES,
            )?;
            anyhow::ensure!(
                response.status() / 100 == 2,
                "Media download failed (HTTP {})",
                response.status()
            );
            let expected_length = usize::try_from(item.content_length()).unwrap_or(usize::MAX);
            anyhow::ensure!(
                response.body().len() == expected_length,
                "Media download size did not match Nintendo's content length"
            );
            anyhow::ensure!(
                response.body().len() <= MAX_MEDIA_DOWNLOAD_BYTES,
                "Media download exceeded the 256 MiB safety limit"
            );
            check_cancelled(&cancelled)?;

            let part = with_appended_suffix(&destination, ".part");
            let _ = fs::remove_file(&part);
            let write_result = (|| -> anyhow::Result<()> {
                let mut file = File::create(&part)?;
                file.write_all(response.body())?;
                file.sync_all()?;
                check_cancelled(&cancelled)?;
                fs::rename(&part, &destination)?;
                Ok(())
            })();
            if let Err(error) = write_result {
                let _ = fs::remove_file(&part);
                return Err(error);
            }

            preserve_capture_timestamp(&destination, timestamp);
            existing.names.insert(filename.to_ascii_lowercase());
            existing.names.insert(prefix.to_ascii_lowercase());
            downloaded += 1;
        }

        Ok(SyncResult::new(media.len(), downloaded))
    }
}

fn validate_media_item(item: &MediaItem) -> anyhow::Result<()> {
    anyhow::ensure!(
        item.content_length() > 0,
        "Nintendo media item has no valid content length"
    );
    anyhow::ensure!(
        item.content_length() <= MAX_MEDIA_DOWNLOAD_BYTES as i64,
        "Nintendo media item exceeds the 256 MiB safety limit"
    );
    let url = url::Url::parse(item.content_uri())?;
    anyhow::ensure!(url.scheme() == "https", "Nintendo media URL must use HTTPS");
    anyhow::ensure!(
        url.username().is_empty() && url.password().is_none(),
        "Nintendo media URL cannot contain credentials"
    );
    anyhow::ensure!(
        url.port_or_known_default() == Some(443),
        "Nintendo media URL must use the standard HTTPS port"
    );
    let host = url
        .host_str()
        .ok_or_else(|| anyhow::anyhow!("Nintendo media URL is missing a hostname"))?;
    anyhow::ensure!(host.contains('.'), "Nintendo media hostname is not public-looking");
    anyhow::ensure!(
        host.parse::<std::net::IpAddr>().is_err(),
        "Nintendo media URL cannot use a numeric IP address"
    );
    anyhow::ensure!(
        host.bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'.' | b'-')),
        "Nintendo media hostname contains invalid characters"
    );
    let lower = host.to_ascii_lowercase();
    let denied_suffixes = [
        "localhost",
        ".localhost",
        ".local",
        ".localdomain",
        ".internal",
        ".lan",
        ".home",
    ];
    anyhow::ensure!(
        !denied_suffixes
            .iter()
            .any(|suffix| lower == suffix.trim_start_matches('.') || lower.ends_with(suffix)),
        "Nintendo media URL resolves to a local hostname namespace"
    );
    Ok(())
}

#[derive(Default)]
struct ExistingAlbumIndex {
    names: HashSet<String>,
    folder_by_timestamp_prefix: HashMap<String, String>,
}

fn index_existing_album<F>(root: &Path, cancelled: &F) -> anyhow::Result<ExistingAlbumIndex>
where
    F: Fn() -> bool,
{
    let mut index = ExistingAlbumIndex::default();
    for entry in WalkDir::new(root)
        .follow_links(false)
        .into_iter()
        .filter_map(Result::ok)
    {
        check_cancelled(cancelled)?;
        if !entry.file_type().is_file() {
            continue;
        }
        let filename = entry.file_name().to_string_lossy().into_owned();
        let lower = filename.to_ascii_lowercase();
        if lower.ends_with(".part") || lower.ends_with(".tmp") {
            continue;
        }
        if entry
            .metadata()
            .map(|metadata| metadata.len() == 0)
            .unwrap_or(true)
        {
            continue;
        }
        let prefix = prefix_from_existing_filename(entry.path());
        index.names.insert(lower);
        index.names.insert(prefix.to_ascii_lowercase());
        if let Some(parent) = entry
            .path()
            .parent()
            .and_then(Path::file_name)
            .and_then(|name| name.to_str())
        {
            index
                .folder_by_timestamp_prefix
                .entry(prefix.to_ascii_lowercase())
                .or_insert_with(|| parent.to_owned());
        }
    }
    Ok(index)
}

fn learn_title_folders(
    media: &[MediaItem],
    existing: &ExistingAlbumIndex,
) -> HashMap<String, String> {
    let mut folders = HashMap::new();
    for item in media {
        if item.title_id().is_empty() {
            continue;
        }
        let prefix = capture_timestamp_prefix(item.timestamp()).to_ascii_lowercase();
        if let Some(folder) = existing.folder_by_timestamp_prefix.get(&prefix) {
            folders.insert(item.title_id().to_ascii_lowercase(), folder.clone());
        }
    }
    folders
}

fn resolve_game_folder(album_directory: &Path, app_name: &str) -> String {
    let default_clean = sanitize_folder_v1(app_name);
    if app_name.trim().is_empty() || !album_directory.is_dir() {
        return default_clean;
    }

    let directories = fs::read_dir(album_directory)
        .ok()
        .into_iter()
        .flatten()
        .filter_map(Result::ok)
        .filter(|entry| entry.file_type().map(|kind| kind.is_dir()).unwrap_or(false))
        .filter_map(|entry| entry.file_name().into_string().ok())
        .map(|name| {
            let normalized = normalize_for_matching_v1(&name);
            (name, normalized)
        })
        .collect::<Vec<_>>();

    let normalized_app = normalize_for_matching_v1(app_name);
    let synonyms: Vec<&str> = GAME_ALIAS_GROUPS
        .iter()
        .find(|group| {
            group
                .iter()
                .any(|alias| normalize_for_matching_v1(alias) == normalized_app)
        })
        .map(|group| group.to_vec())
        .unwrap_or_else(|| vec![app_name]);

    for synonym in synonyms {
        let clean = sanitize_folder_v1(synonym);
        let normalized = normalize_for_matching_v1(synonym);
        if let Some((name, _)) = directories
            .iter()
            .find(|(name, _)| equal_v1_name(name, &clean))
        {
            return name.clone();
        }
        if let Some((name, _)) = directories
            .iter()
            .find(|(_, candidate)| candidate == &normalized)
        {
            return name.clone();
        }
    }

    if normalized_app.len() >= 6 {
        if let Some((name, _)) = directories.iter().find(|(_, candidate)| {
            candidate.len() >= 6
                && (candidate.contains(&normalized_app) || normalized_app.contains(candidate))
        }) {
            return name.clone();
        }
    }

    default_clean
}

fn sanitize_folder_v1(value: &str) -> String {
    if value.trim().is_empty() {
        return "Other".to_owned();
    }
    let mut clean = String::with_capacity(value.len());
    for character in value.chars() {
        if character.is_control() {
            continue;
        }
        if is_unicode_dash_or_hyphen(character) {
            clean.push('-');
            continue;
        }
        if is_invalid_filename_character(character) {
            continue;
        }
        clean.push(character);
    }

    let mut collapsed = String::with_capacity(clean.len());
    let mut previous_space = false;
    for character in clean.chars() {
        if character == ' ' {
            if !previous_space && !collapsed.is_empty() {
                collapsed.push(' ');
                previous_space = true;
            }
        } else {
            collapsed.push(character);
            previous_space = false;
        }
    }
    while matches!(collapsed.chars().last(), Some(' ' | '.')) {
        collapsed.pop();
    }
    let result = collapsed.trim().to_owned();
    if result.is_empty() { "Other".to_owned() } else { result }
}

fn is_invalid_filename_character(character: char) -> bool {
    matches!(
        character,
        '<' | '>' | ':' | '"' | '/' | '\\' | '|' | '?' | '*'
            | '\u{FF1C}' | '\u{FF1E}' | '\u{FF1A}' | '\u{FF02}' | '\u{FF0F}'
            | '\u{FF3C}' | '\u{FF5C}' | '\u{FF1F}' | '\u{FF0A}'
    )
}

fn is_unicode_dash_or_hyphen(character: char) -> bool {
    matches!(
        character,
        '\u{2010}' | '\u{2011}' | '\u{2012}' | '\u{2013}' | '\u{2014}'
            | '\u{2015}' | '\u{2212}' | '\u{FE58}' | '\u{FE63}' | '\u{FF0D}'
    )
}

fn normalize_for_matching_v1(value: &str) -> String {
    let mut output = String::with_capacity(value.len());
    for character in value.nfd() {
        if is_combining_mark(character) {
            continue;
        }
        if character.is_ascii() {
            if character.is_ascii_alphanumeric() {
                output.push(character.to_ascii_lowercase());
            }
        } else {
            output.push(character);
        }
    }
    output
}

fn equal_v1_name(left: &str, right: &str) -> bool {
    left.eq_ignore_ascii_case(right)
}

fn capture_timestamp_prefix(timestamp: i64) -> String {
    use chrono::{Local, TimeZone as _};
    let supplied = if timestamp > 10_000_000_000 {
        timestamp / 1000
    } else {
        timestamp
    };
    let seconds = if supplied <= 0 {
        Local::now().timestamp()
    } else {
        supplied
    };
    let local = Local
        .timestamp_opt(seconds, 0)
        .single()
        .or_else(|| Local.timestamp_opt(seconds, 0).earliest())
        .unwrap_or_else(Local::now);
    local.format("%Y%m%d%H%M%S00").to_string()
}

fn prefix_from_existing_filename(path: &Path) -> String {
    let mut prefix = path
        .file_stem()
        .and_then(|name| name.to_str())
        .unwrap_or_default()
        .to_owned();
    if prefix.ends_with("_c") {
        prefix.truncate(prefix.len().saturating_sub(2));
    } else if prefix.ends_with("-00") {
        prefix.truncate(prefix.len().saturating_sub(3));
    }
    prefix
}

fn preserve_capture_timestamp(path: &Path, timestamp: i64) {
    if timestamp <= 0 {
        return;
    }
    let seconds = if timestamp > 10_000_000_000 {
        timestamp / 1000
    } else {
        timestamp
    };
    let _ = set_file_mtime(path, FileTime::from_unix_time(seconds, 0));
}

fn with_appended_suffix(path: &Path, suffix: &str) -> PathBuf {
    let mut value = path.as_os_str().to_os_string();
    value.push(suffix);
    PathBuf::from(value)
}

fn check_cancelled<F>(cancelled: &F) -> anyhow::Result<()>
where
    F: Fn() -> bool,
{
    anyhow::ensure!(!cancelled(), "sync cancelled");
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::{normalize_for_matching_v1, sanitize_folder_v1, validate_media_item};
    use crate::model::MediaItem;

    #[test]
    fn folder_matching_retains_v1_semantics() {
        assert_eq!(normalize_for_matching_v1("Pokémon: Violet"), "pokemonviolet");
        assert_eq!(sanitize_folder_v1("Bad:/Name*"), "BadName");
        assert_eq!(sanitize_folder_v1("A—B"), "A-B");
    }

    #[test]
    fn media_validation_rejects_local_hostnames() {
        let item: MediaItem = serde_json::from_value(serde_json::json!({
            "contentUri": "https://localhost/x",
            "contentLength": 10
        }))
        .expect("deserialize media");
        assert!(validate_media_item(&item).is_err());
    }
}
